### ISSUE TITLE
$1000 - When you edit a comment, the composer stays hidden on other chats - reported by @thesahindia

### ISSUE THREAD

**Issue by mvtglobally** [Expensify/App#8031]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:

1. Open :expensify: on Mobile
2. Navigate to a chat
3. Edit a comment
4. Navigate to a different chat

## Expected Result:
Composer should be visible on other chats or user shouldn't be able to navigate to different chat until he cancels/Save the changes

## Actual Result:
Composer is hidden for every chat

## Workaround:
unknown 

## Platform:

Where is this issue occurring?

- iOS
- Android
- Mobile Web

**Version Number:** 1.1.41-0
**Reproducible in staging?:** Y 
**Reproducible in production?:** Y 
**Email or phone of affected tester (no customers):** 
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation 

https://user-images.githubusercontent.com/43995119/157183657-c4c70cbf-047d-438e-9e82-7f90cc0dc139.mov

**Expensify/Expensify Issue URL:** 
**Issue reported by:** @thesahindia
**Slack conversation:** https://expensify.slack.com/archives/C01GTK53T8Q/p1645699620022559

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by MelvinBot** [comment]
Triggered auto assignment to @laurenreidexpensify (`AutoAssignerTriage`), see https://stackoverflow.com/c/expensify/questions/4749 for more details.

**Comment by laurenreidexpensify** [comment]
@thesahindia I'm not sure I follow full here - what do you mean by "edit a comment". Do you mean editing a message that has already been sent?

**Comment by MelvinBot** [comment]
Triggered auto assignment to @robertjchen (`Engineering`), see https://stackoverflow.com/c/expensify/questions/4319 for more details.

**Comment by robertjchen** [comment]
> Composer should be visible on other chats or user shouldn't be able to navigate to different chat until he cancels/Save the changes

Got it! So to clarify my understanding, what should be the proper behavior here? 

At the moment, the composer stays open after switching to another chat and switching back. I don't see it as a major issue, but I do see three options to make it "better":
1. Automatically cancel the edit and hide the composer when switching a chat (Slack does this)
2. Composer should remain visible on other chats (this might be quite confusing)
3. Prompt the user to save/cancel the chat (e.g. highlight composer box in red and prevent user from switching, a dialog, etc.)

I see 1 as the path forward here- I'm open to any other thoughts!

**Comment by thesahindia** [comment]
I like the 2nd option. 1st option is good but there is a drawback. If the user is in the middle of editing and navigates to some other chat then the changes he made will get discarded.

**Comment by robertjchen** [comment]
Got it, that makes sense- it should mirror the behavior on the desktop version (which is working properly, just not on mobile). This is clear to me now, apologies for any initial confusion!

So, the report/proposal here is to address the issue where the composer disappears when switching to other chats on mobile 👍  Thanks again for the report- going to add the `External` label for the next step of the process.

**Comment by MelvinBot** [comment]
Triggered auto assignment to @MitchExpensify (`External`), see https://stackoverflow.com/c/expensify/questions/8582 for more details.

**Comment by MitchExpensify** [comment]
[Upwork Job](https://www.upwork.com/jobs/~01e3f91f64336faaee)

**Comment by MelvinBot** [comment]
Triggered auto assignment to Contributor-plus team member for initial proposal review - @Santhosh-Sellavel (`Exported`)

**Comment by MelvinBot** [comment]
Triggered auto assignment to @Beamanator (`Exported`), see https://stackoverflow.com/c/expensify/questions/7972 for more details.

**Comment by Beamanator** [comment]
Hmm unable to reproduce on mobile web (chrome), latest version - I'll try android

**Comment by Santhosh-Sellavel** [comment]
@Beamanator Able to reproduce in mWeb

https://user-images.githubusercontent.com/85645967/158416332-14777a7e-6080-41fb-a308-b6b3c6043e4c.MOV

**Comment by Beamanator** [comment]
@Santhosh-Sellavel thanks for the vid, what platform & version are you on?

I just tried on Android, in Chrome browser, v1.1.42-6 and still couldn't reproduce 😅

**Comment by Beamanator** [comment]
Waiting on proposals 👍

**Comment by Beamanator** [comment]
I guess the interesting thing is:
1. In web (and Desktop, I assume), when you edit a comment the ReportActionCompose is still showing
2. In mobile (and mobile web), when you edit a comment the ReportActionCompose disappears even for the chat you're editing

I am guessing this is due to a recent PR, let's see if we can nail that down

**Comment by thesahindia** [comment]
> 1. In web (and Desktop, I assume), when you edit a comment the ReportActionCompose is still showing
> 2. In mobile (and mobile web), when you edit a comment the ReportActionCompose disappears even for the chat you're editing

@Beamanator, I think this is expected, info about it is in https://github.com/Expensify/App/issues/7265

**Comment by Beamanator** [comment]
Aha thank you @thesahindia ! 

@parasharrajat could this issue be a regression of your PR https://github.com/Expensify/App/pull/7390?

**Comment by parasharrajat** [comment]
Yeah, this is a side-effect from #7390. But I just enabled the flag which was missing from the code. This means this functionality wasn't developed correctly and it had this bug. 

Let me check this issue quickly...

**Comment by parasharrajat** [comment]
### Proposal 

1. In ReportScreen's `componentWillUnmount` call `toggleReportActionComposeView(true, true)`. It will remount the Composer when user closes the report.

The best behavior would be option 1. Where we remount the composer on each report change.

**Comment by Beamanator** [comment]
If we do that, what if the user:
1. Starts editing a comment (so the ReportActionCompose is hidden in that chat)
2. Navigates to a new chat (now ReportActionCompose should show)
3. Navigates back to the original chat

In this case wouldn't ReportActionCompose be showing, along with the comment being edited?

**Comment by parasharrajat** [comment]
Yeah. It will but as soon as the user clicks on the message edit box, the composer hides which is normally happening when you edit a message.

**Comment by Beamanator** [comment]
@parasharrajat do you mind testing it and throwing in a short vid to explain what you mean? If I'm understanding correctly the `ReportActionCompose` will still be showing (in step 3 above) but it will hide if the user clicks in the comment-edit box, which seems like it would result in a weird user experience, maybe I don't understand completely?

**Comment by parasharrajat** [comment]
Your understanding is correct. 

> which seems like it would result in a weird user experience

If you feel so then we have to think about an alternative strategy. But I think it is the best fit. On Mobile, we don't focus anything when the report screens loads. There can be multiple draft messages. 

1. Either we remove this feature completely. 
2. Or find something else.

**Comment by Beamanator** [comment]
Ok so now I'm thinking this does actually make sense to keep around on mobile because:
1. When you originally click "Edit comment", you won't be sending a message so it makes sense to hide `ReportActionCompose`
2. When you cancel or save your edit, `ReportActionCompose` comes back (since now you'll be thinking about sending a message)
3. When you navigate away & come back, you may want to send a message or continue editing, so if you click back in the edit comment box we should hide the `ReportActionCompose` to make it consistent

I like the idea now, sorry for flip-flopping a bit. @parasharrajat Let's go with your proposal [here](https://github.com/Expensify/App/issues/8031#issuecomment-1069077890)

**Comment by MelvinBot** [comment]
📣 @parasharrajat You have been assigned to this job by @Beamanator!
Please apply to this job in Upwork and leave a comment on the Github issue letting us know when we can expect a PR to be ready for review 🧑‍💻
Keep in mind: [Code of Conduct](https://github.com/Expensify/App/blob/main/CODE_OF_CONDUCT.md) | [Contributing](https://github.com/Expensify/App/blob/main/CONTRIBUTING.md) 📖

**Comment by parasharrajat** [comment]
No worries. I appreciate the clarifying questions. It is actually better to discuss it so we are on the same boat.

**Comment by MitchExpensify** [comment]
Invited to job @parasharrajat

**Comment by parasharrajat** [comment]
@Beamanator question? What should happen when?
1. user was editing a message and the composer is hidden.
2. He goes back to LHN.
3. Then reopen the same report.
4. Should the composer be shown or not? 

So far, I am only forcing the composer on report change.

**Comment by parasharrajat** [comment]
Asked here https://expensify.slack.com/archives/C01GTK53T8Q/p1647881926785649

**Comment by Beamanator** [comment]
@parasharrajat I think we got the answers we need in Slack, do you have any questions before moving forward with a PR?

**Comment by parasharrajat** [comment]
I will update the PR by monday.

**Comment by parasharrajat** [comment]
> Thanks ... I like that idea too, I agree it’s not necessary to keep the edit box open when navigating to LHN and back :thumbsup:

@Beamanator Do I need to add this as well?

**Comment by Beamanator** [comment]
@parasharrajat Yes we should include the decision made in Slack. Would you mind restating exactly what you plan to do before moving forward with this?

**Comment by parasharrajat** [comment]
Ok, I found that if we want to close the edit box while going back to LHN, we have to lose the draft messages as well. 

Edit box depends on the presence of draft message directly https://github.com/Expensify/App/blob/1ff6625a19f56f9d636edc09a1d2f76d5f7a893c/src/pages/home/report/ReportActionItem.js#L139

**Comment by Beamanator** [comment]
Ok, personally I think that's ok because it's just a draft **edit**

**Comment by parasharrajat** [comment]
Ok, I will update the PR.

**Comment by parasharrajat** [comment]
@Beamanator  Ok, So there are two ways, I can think of. 
- I use `withReportActionsDrafts` HOC on ReportScreen to get all the reportActionsID with keys. Then use WithDrawerState to set those drafts off when the drawer is opened.
- Secondly, I can directly use `WithDrawerState` on the ReportActionItem and reject the draft when the drawer is Opened.

**Comment by Beamanator** [comment]
Ok so just to make sure I'm following correctly:
1. `this.props.draftMessage` (in `ReportActionItem.js`) is where we're storing the draft of an edited message (not the draft of a new message)
2. You're thinking we can just kill a report's draft (probably using `Report.saveReportActionDraft`, storing `null`) when the LHN opens up

If this is correct, I'm thinking the second option may be more efficient since we only have to deal with 1 report's draft at a time, what do you think?

**Comment by parasharrajat** [comment]
Yup, you are correct. But a report can have multiple drafts for actions.

**Comment by Beamanator** [comment]
Ok and that's why in your sample code you're looping through all drafts of a specific `reportID`, which I bet you'd have to do in either case, thanks for reminding me about that 👍 

A few more questions related to this change we're thinking about making:
1. Now that we're planning to get rid of message-edits once a user navigates back to the LHN, do you think it even makes sense to store these in Onyx anymore? A.k.a. if we don't care about message-edit drafts when we switch to a new report, wouldn't we also not care about saving the drafts if the user closes the app, for example?
2. On large screen devices we won't "open the LHN" ever so to make behavior consistent would we want to kill message-edit drafts if the user changes reports? Or do we want to :donothing: for big screen devices? I'm leaning toward :donothing:

**Comment by parasharrajat** [comment]
Yeah, good points. 

I have a second thought about this change. We are basically removing the drafts feature from the app. Do you think we should take the opinion of a few more people? Let me know if this is the final decision, I will look into your questions as well.

**Comment by Beamanator** [comment]
Yeah I was thinking something similar 👍 I agree we should get more opinions, if you could continue the same chain as before but re-send to the channel & maybe tag contributor-management-engineering as well that would be very helpful

**Comment by MitchExpensify** [comment]
[New Upwork Job](https://www.upwork.com/jobs/~0127d9aa2ccfeabd34) (previous one closed)

Price doubled!

**Comment by Beamanator** [comment]
Ping'd CMEs in internal channel to try to get more eyes 👁️

**Comment by neil-marcellini** [comment]
> We are basically removing the drafts feature from the app.

I don't think we should do this. If I start editing a message on a chat, and then I need to go reference something from another chat, I want to be able to come back to my draft and continue editing. If I have been editing a message and I go to another chat I should also be able to send messages there. That should be possible right? Is there another case I'm not thinking of?

**Comment by parasharrajat** [comment]
> If I have been editing a message and I go to another chat I should also be able to send messages there. That should be possible right?

Yeah, that is possible.

**Comment by neil-marcellini** [comment]
>What should happen when?
> 
> 1. user was editing a message and the composer is hidden.
> 2. He goes back to LHN.
> 3. Then reopen the same report.
> 4. Should the composer be shown or not?

If you do this on web / desktop then the composer will be shown, so I think it should match on mobile. In my opinion, we should let the user have the option to continue editing or to send another message. Again that behavior would match desktop. To make that happen I think we should allow the user to unfocus the composer (and dismiss the keyboard) if they click outside of it on mobile. Maybe that's a separate issue.

**Comment by Beamanator** [comment]
Left a big comment w/ multiple ideas & ping'd a few more people in Slack: https://expensify.slack.com/archives/C01GTK53T8Q/p1651495589126309?thread_ts=1647881926.785649&cid=C01GTK53T8Q

**Comment by frenkield** [comment]
So sorry, I'm new to Expensify and I don't understand what the issue is here. Where in the attached video is something incorrect?

**Comment by MitchExpensify** [comment]
Doubled!

**Comment by Beamanator** [comment]
Ping'd the slack channel again to see if anyone can help come to a final decision :D

**Comment by parasharrajat** [comment]
Bump! Any update?

**Comment by Beamanator** [comment]
I'm getting fewer and fewer responses to my posts in slack 😅 I don't feel comfortable moving forward till we have more feedback on the thought so I will tag CMEs and try to get more thoughts

@MitchExpensify I think you can hold off doubling this time please :D Since we don't have a solid way forward yet

**Comment by marcaaron** [comment]
Heads up I'm working on this internally in a [duplicate issue here](https://github.com/Expensify/App/issues/9016) and found a pretty simple solution.

**Comment by marcaaron** [comment]
I think the correct behavior is that we should hide and show the "composer" only when the edit comment field is focused or blurred.

**Comment by marcaaron** [comment]
Has anyone been hired? If not, I think we can just close this one (and the Upwork job) and have @parasharrajat and @Beamanator test my PR linked to the duplicate issue.

**Comment by Beamanator** [comment]
@marcaaron I ended up hiring @parasharrajat here: https://github.com/Expensify/App/issues/8031#issuecomment-1073543793

After that, we've been discussing what to do in some other situations (see Slack thread: https://expensify.slack.com/archives/C01GTK53T8Q/p1652268683129369?thread_ts=1647881926.785649&cid=C01GTK53T8Q)

And this has been delayed due to lack of responses, which is why I just brought this to an internal slack channel as well

**Comment by Beamanator** [comment]
If you already have a PR up, maybe we can pay @parasharrajat to help test & review it? (I'll help as well)

**Comment by parasharrajat** [comment]
What is decided here? Should I close my PR?

**Comment by marcaaron** [comment]
Yes, want to help us review this solution instead --> https://github.com/Expensify/App/pull/9060

**Comment by marcaaron** [comment]
Added you guys there.

**Comment by Beamanator** [comment]
Not overdue, PR testing and reviewing in progress

**Comment by Beamanator** [comment]
I believe this can be closed b/c the fix was implemented in this PR: https://github.com/Expensify/App/pull/9060

**Comment by MitchExpensify** [comment]
Agree!

**Comment by parasharrajat** [comment]
Note: I was hired for this issue [here](https://github.com/Expensify/App/issues/8031#issuecomment-1073543793) and It was decided that this issue will be solve internally and [I will review it as C+](https://github.com/Expensify/App/issues/8031#issuecomment-1130571296). Also, I closed my PR for the same.

**Comment by parasharrajat** [comment]
Yeah, you are correct. Thanks.

**Comment by thesahindia** [comment]
@MitchExpensify, the reporting compensation is also pending.

**Comment by MitchExpensify** [comment]
Thanks for the heads up team! I have made a note to issue payment tomorrow as that will be 7 days since the fix was deployed [here](https://github.com/Expensify/App/pull/9060)

**Comment by MitchExpensify** [comment]
Payments made and contract ended - Thanks @parasharrajat & @thesahindia!
